// @remotion/studio-shared/dist/schema-field-info.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getEffectFieldsToShow = exports.getFieldsToShow = exports.SCHEMA_FIELD_ROW_HEIGHT = void 0;
const remotion_1 = require("remotion");
const no_react_1 = require("remotion/no-react");
exports.SCHEMA_FIELD_ROW_HEIGHT = 22;
const SUPPORTED_SCHEMA_TYPES = [
    'number',
    'boolean',
    'rotation',
    'translate',
    'color',
    'enum',
    'hidden',
];
const getFieldsToShow = ({ getDragOverrides, codeValues, nodePath, schema, currentRuntimeValueDotNotation, }) => {
    const { merged: valuesDotNotation } = remotion_1.Internals.computeEffectiveSchemaValuesDotNotation({
        schema,
        currentValue: currentRuntimeValueDotNotation,
        overrideValues: getDragOverrides(nodePath),
        propStatus: remotion_1.Internals.getCodeValuesCtx(codeValues, nodePath),
    });
    const activeSchema = remotion_1.Internals.flattenActiveSchema(schema, (key) => valuesDotNotation[key]);
    return Object.entries(activeSchema)
        .map(([key, fieldSchema]) => {
        const typeName = fieldSchema.type;
        if (SUPPORTED_SCHEMA_TYPES.indexOf(typeName) === -1) {
            throw new Error(`Unsupported field type: ${typeName}`);
        }
        if (typeName === 'hidden') {
            return null;
        }
        // `hidden` is represented as the eye/speaker icon on the timeline track,
        // so we don't render it as a regular field in the expanded section.
        if (key === 'hidden') {
            return null;
        }
        return {
            kind: 'sequence-field',
            key,
            description: fieldSchema.description,
            typeName,
            rowHeight: exports.SCHEMA_FIELD_ROW_HEIGHT,
            fieldSchema,
        };
    })
        .filter(no_react_1.NoReactInternals.truthy);
};
exports.getFieldsToShow = getFieldsToShow;
const getEffectFieldsToShow = (effect, effectIndex) => {
    return Object.entries(effect.schema)
        .map(([key, fieldSchema]) => {
        const typeName = fieldSchema.type;
        if (typeName === 'hidden') {
            return null;
        }
        // `disabled` is represented as the eye icon on the effect timeline row,
        // so we don't render it as a regular field in the expanded section.
        if (key === 'disabled') {
            return null;
        }
        if (SUPPORTED_SCHEMA_TYPES.indexOf(typeName) === -1) {
            throw new Error(`Unsupported field type: ${typeName}`);
        }
        return {
            kind: 'effect-field',
            key,
            description: fieldSchema.description,
            typeName,
            rowHeight: exports.SCHEMA_FIELD_ROW_HEIGHT,
            fieldSchema,
            effectSchema: effect.schema,
            effectIndex,
        };
    })
        .filter(no_react_1.NoReactInternals.truthy);
};
exports.getEffectFieldsToShow = getEffectFieldsToShow;
