// @remotion/studio/dist/components/Timeline/save-effect-prop.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.saveEffectProp = void 0;
const studio_shared_1 = require("@remotion/studio-shared");
const call_api_1 = require("../call-api");
const save_prop_queue_1 = require("./save-prop-queue");
const saveEffectProp = ({ fileName, nodePath, effectIndex, fieldKey, value, defaultValue, schema, setCodeValues, }) => {
    return (0, save_prop_queue_1.enqueueSavePropChange)({
        nodePath,
        setCodeValues,
        applyOptimistic: (prev) => (0, studio_shared_1.optimisticUpdateForEffectCodeValues)({
            previous: prev,
            effectIndex,
            fieldKey,
            value,
            schema,
        }),
        apiCall: () => (0, call_api_1.callApi)('/api/save-effect-props', {
            fileName,
            sequenceNodePath: nodePath,
            effectIndex,
            key: fieldKey,
            value: JSON.stringify(value),
            defaultValue,
            schema,
        }),
        mergeServerResponse: (prev, data) => {
            if (!prev.canUpdate) {
                return prev;
            }
            const idx = prev.effects.findIndex((e) => e.effectIndex === effectIndex);
            if (idx === -1) {
                return { ...prev, effects: [...prev.effects, data] };
            }
            const nextEffects = [...prev.effects];
            nextEffects[idx] = data;
            return { ...prev, effects: nextEffects };
        },
        errorLabel: 'Could not save effect prop',
    });
};
exports.saveEffectProp = saveEffectProp;
