// @remotion/studio/dist/components/Timeline/TimelineColorField.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineColorField = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const no_react_1 = require("remotion/no-react");
const colors_1 = require("../../helpers/colors");
const eyedropper_1 = require("../../icons/eyedropper");
const z_index_1 = require("../../state/z-index");
const SWATCH_WIDTH = 20;
const SWATCH_HEIGHT = 15;
const containerStyle = {
    display: 'flex',
    alignItems: 'center',
    gap: 3,
};
const swatchWrapperBase = {
    position: 'relative',
    width: SWATCH_WIDTH,
    height: SWATCH_HEIGHT,
    display: 'inline-block',
    borderRadius: 3,
    overflow: 'hidden',
    cursor: 'pointer',
    borderStyle: 'solid',
    borderWidth: 1,
};
const hiddenInputStyle = {
    position: 'absolute',
    inset: 0,
    width: '100%',
    height: '100%',
    opacity: 0,
    cursor: 'pointer',
    border: 'none',
    padding: 0,
    margin: 0,
};
const swatchFillStyle = {
    width: '100%',
    height: '100%',
};
const eyedropperButtonBase = {
    background: 'transparent',
    border: 'none',
    padding: 0,
    margin: 0,
    cursor: 'pointer',
    display: 'inline-flex',
    alignItems: 'center',
    justifyContent: 'center',
    width: 20,
    height: 20,
    color: 'rgba(255, 255, 255, 0.7)',
};
const eyedropperIconStyle = {
    width: 16,
    height: 16,
};
// Normalizes any color string the user provided (e.g. `red`, `rgb(...)`, `#fff`)
// into a `#rrggbb` string that `<input type="color">` accepts.
const toHex = (value) => {
    try {
        const argb = no_react_1.NoReactInternals.processColor(value);
        const r = (argb >>> 16) & 0xff;
        const g = (argb >>> 8) & 0xff;
        const b = argb & 0xff;
        return `#${r.toString(16).padStart(2, '0')}${g
            .toString(16)
            .padStart(2, '0')}${b.toString(16).padStart(2, '0')}`;
    }
    catch (_a) {
        return '#000000';
    }
};
const hasEyeDropper = () => typeof window !== 'undefined' && 'EyeDropper' in window;
const TimelineColorField = ({ field, effectiveValue, propStatus, onSave, onDragValueChange, onDragEnd, }) => {
    const inputRef = (0, react_1.useRef)(null);
    const [isHovered, setIsHovered] = (0, react_1.useState)(false);
    const [isFocused, setIsFocused] = (0, react_1.useState)(false);
    const { tabIndex } = (0, z_index_1.useZIndex)();
    const commitTimeoutRef = (0, react_1.useRef)(null);
    const pendingCommitRef = (0, react_1.useRef)(null);
    // `<input type="color">` doesn't fire an event when dismissed; debounce
    // commits and flush any pending commit on unmount so we never lose the
    // final value.
    (0, react_1.useEffect)(() => {
        return () => {
            if (commitTimeoutRef.current) {
                clearTimeout(commitTimeoutRef.current);
            }
            if (pendingCommitRef.current) {
                pendingCommitRef.current();
            }
        };
    }, []);
    const currentValue = typeof effectiveValue === 'string'
        ? effectiveValue
        : field.fieldSchema.type === 'color'
            ? field.fieldSchema.default
            : '';
    const hexValue = (0, react_1.useMemo)(() => toHex(currentValue), [currentValue]);
    const onColorChange = (0, react_1.useCallback)((e) => {
        const newValue = e.target.value;
        if (!propStatus.canUpdate) {
            return;
        }
        onDragValueChange(newValue);
        if (commitTimeoutRef.current) {
            clearTimeout(commitTimeoutRef.current);
        }
        const commit = () => {
            pendingCommitRef.current = null;
            if (propStatus.canUpdate && newValue !== propStatus.codeValue) {
                onSave(newValue);
            }
            onDragEnd();
        };
        pendingCommitRef.current = commit;
        commitTimeoutRef.current = setTimeout(() => {
            commitTimeoutRef.current = null;
            commit();
        }, 500);
    }, [onSave, onDragValueChange, onDragEnd, propStatus]);
    const onPickColor = (0, react_1.useCallback)(() => {
        // `EyeDropper` is a Chromium-only API; it's feature-detected at render
        // time so this only runs in supported browsers.
        const EyeDropperCtor = typeof window !== 'undefined'
            ? window.EyeDropper
            : undefined;
        if (!EyeDropperCtor) {
            return;
        }
        const eyeDropper = new EyeDropperCtor();
        eyeDropper
            .open()
            .then((result) => {
            if (propStatus.canUpdate && result.sRGBHex !== propStatus.codeValue) {
                onSave(result.sRGBHex);
            }
        })
            .catch(() => {
            // User aborted or picker failed; safe to ignore.
        });
    }, [onSave, propStatus]);
    const swatchWrapperStyle = (0, react_1.useMemo)(() => {
        return {
            ...swatchWrapperBase,
            borderColor: isHovered || isFocused
                ? colors_1.INPUT_BORDER_COLOR_HOVERED
                : colors_1.INPUT_BORDER_COLOR_UNHOVERED,
            cursor: propStatus.canUpdate ? 'pointer' : 'not-allowed',
            marginLeft: 5,
        };
    }, [isFocused, isHovered, propStatus.canUpdate]);
    const swatchFill = (0, react_1.useMemo)(() => {
        return {
            ...swatchFillStyle,
            backgroundColor: currentValue || hexValue,
            position: 'absolute',
            display: 'block',
        };
    }, [currentValue, hexValue]);
    const onMouseEnter = (0, react_1.useCallback)(() => setIsHovered(true), []);
    const onMouseLeave = (0, react_1.useCallback)(() => setIsHovered(false), []);
    const onFocus = (0, react_1.useCallback)(() => setIsFocused(true), []);
    const onBlur = (0, react_1.useCallback)(() => setIsFocused(false), []);
    const onSwatchClick = (0, react_1.useCallback)(() => {
        var _a;
        if (!propStatus.canUpdate) {
            return;
        }
        (_a = inputRef.current) === null || _a === void 0 ? void 0 : _a.click();
    }, [propStatus.canUpdate]);
    const showEyeDropper = hasEyeDropper();
    return (jsx_runtime_1.jsxs("span", { style: containerStyle, children: [
            jsx_runtime_1.jsxs("span", { style: swatchWrapperStyle, onMouseEnter: onMouseEnter, onMouseLeave: onMouseLeave, onClick: onSwatchClick, title: currentValue, children: [
                    jsx_runtime_1.jsx("span", { style: swatchFill }), jsx_runtime_1.jsx("input", { ref: inputRef, type: "color", value: hexValue, onChange: onColorChange, onFocus: onFocus, onBlur: onBlur, disabled: !propStatus.canUpdate, name: field.key, tabIndex: tabIndex, style: hiddenInputStyle })
                ] }), showEyeDropper ? (jsx_runtime_1.jsx("button", { type: "button", onClick: onPickColor, disabled: !propStatus.canUpdate, style: eyedropperButtonBase, tabIndex: tabIndex, title: "Pick color from screen", "aria-label": "Pick color from screen", children: jsx_runtime_1.jsx(eyedropper_1.EyedropperIcon, { style: eyedropperIconStyle, color: "rgba(255, 255, 255, 0.7)" }) })) : null] }));
};
exports.TimelineColorField = TimelineColorField;
