// @remotion/studio/dist/components/Timeline/TimelineExpandedRow.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineExpandedRow = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const timeline_layout_1 = require("../../helpers/timeline-layout");
const Padder_1 = require("./Padder");
const TimelineEffectFieldRow_1 = require("./TimelineEffectFieldRow");
const TimelineEffectGroupRow_1 = require("./TimelineEffectGroupRow");
const TimelineExpandArrowButton_1 = require("./TimelineExpandArrowButton");
const TimelineFieldRow_1 = require("./TimelineFieldRow");
const TimelineListItem_1 = require("./TimelineListItem");
const groupRowBase = {
    height: timeline_layout_1.TREE_GROUP_ROW_HEIGHT,
    display: 'flex',
    alignItems: 'center',
    paddingRight: timeline_layout_1.EXPANDED_SECTION_PADDING_RIGHT,
};
const rowLabel = {
    fontSize: 12,
    color: 'rgba(255, 255, 255, 0.8)',
    userSelect: 'none',
};
const labelOnlyRowBase = {
    display: 'flex',
    alignItems: 'center',
    paddingRight: timeline_layout_1.EXPANDED_SECTION_PADDING_RIGHT,
};
const TimelineExpandedRow = ({ node, depth, nestedDepth, getIsExpanded, toggleTrack, validatedLocation, nodePath, schema, }) => {
    const paddingLeft = timeline_layout_1.EXPANDED_SECTION_PADDING_LEFT + depth * TimelineListItem_1.INDENT;
    const groupStyle = (0, react_1.useMemo)(() => ({ ...groupRowBase, paddingLeft }), [paddingLeft]);
    const labelOnlyStyle = (0, react_1.useMemo)(() => ({
        ...labelOnlyRowBase,
        height: (0, timeline_layout_1.getTreeRowHeight)(node),
        paddingLeft,
    }), [node, paddingLeft]);
    if (node.kind === 'group') {
        if (node.effectInfo) {
            return (jsx_runtime_1.jsx(TimelineEffectGroupRow_1.TimelineEffectGroupRow, { label: node.label, nodePathInfo: node.nodePathInfo, effectIndex: node.effectInfo.effectIndex, effectSchema: node.effectInfo.effectSchema, nodePath: nodePath, validatedLocation: validatedLocation, nestedDepth: nestedDepth, style: groupStyle, getIsExpanded: getIsExpanded, toggleTrack: toggleTrack }));
        }
        const isExpanded = getIsExpanded(node.nodePathInfo);
        return (jsx_runtime_1.jsxs("div", { style: groupStyle, children: [
                jsx_runtime_1.jsx(Padder_1.Padder, { depth: nestedDepth + 1 }), jsx_runtime_1.jsx(TimelineExpandArrowButton_1.TimelineExpandArrowButton, { isExpanded: isExpanded, onClick: () => toggleTrack(node.nodePathInfo), label: `${node.label} section`, disabled: false }), jsx_runtime_1.jsx("span", { style: rowLabel, children: node.label })
            ] }));
    }
    if (node.field) {
        if (node.field.kind === 'effect-field') {
            return (jsx_runtime_1.jsx(TimelineEffectFieldRow_1.TimelineEffectFieldRow, { field: node.field, validatedLocation: validatedLocation, paddingLeft: paddingLeft, nestedDepth: nestedDepth, nodePath: nodePath }));
        }
        if (node.field.kind === 'sequence-field') {
            return (jsx_runtime_1.jsx(TimelineFieldRow_1.TimelineFieldRow, { field: node.field, validatedLocation: validatedLocation, paddingLeft: paddingLeft, nestedDepth: nestedDepth, nodePath: nodePath, schema: schema }));
        }
        throw new Error('Unexpected field kind: ' + JSON.stringify(node.field));
    }
    return (jsx_runtime_1.jsxs("div", { style: labelOnlyStyle, children: [
            jsx_runtime_1.jsx(Padder_1.Padder, { depth: nestedDepth + 1 }), jsx_runtime_1.jsx("span", { style: rowLabel, children: node.label })
        ] }));
};
exports.TimelineExpandedRow = TimelineExpandedRow;
