// @remotion/studio/dist/components/Timeline/TimelineEffectGroupRow.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineEffectGroupRow = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const remotion_1 = require("remotion");
const Padder_1 = require("./Padder");
const save_effect_prop_1 = require("./save-effect-prop");
const TimelineExpandArrowButton_1 = require("./TimelineExpandArrowButton");
const TimelineLayerEye_1 = require("./TimelineLayerEye");
const groupRowBase = {
    display: 'flex',
    alignItems: 'center',
};
const rowLabel = {
    fontSize: 12,
    color: 'rgba(255, 255, 255, 0.8)',
    userSelect: 'none',
};
const TimelineEffectGroupRow = ({ label, nodePathInfo, effectIndex, effectSchema, nodePath, validatedLocation, nestedDepth, style, getIsExpanded, toggleTrack, }) => {
    var _a;
    var _b;
    const { codeValues } = (0, react_1.useContext)(remotion_1.Internals.VisualModeCodeValuesContext);
    const { setCodeValues } = (0, react_1.useContext)(remotion_1.Internals.VisualModeSettersContext);
    const effectStatus = (0, react_1.useMemo)(() => remotion_1.Internals.getEffectCodeValuesCtx({
        codeValues,
        nodePath,
        effectIndex,
    }), [codeValues, nodePath, effectIndex]);
    const disabledStatus = effectStatus.type === 'can-update-effect'
        ? ((_b = (_a = effectStatus.props) === null || _a === void 0 ? void 0 : _a.disabled) !== null && _b !== void 0 ? _b : null)
        : null;
    const isDisabled = (0, react_1.useMemo)(() => {
        if (disabledStatus && disabledStatus.canUpdate) {
            return Boolean(disabledStatus.codeValue);
        }
        return false;
    }, [disabledStatus]);
    const canToggle = disabledStatus !== null && disabledStatus.canUpdate;
    const onToggle = (0, react_1.useCallback)((type) => {
        if (!canToggle) {
            return;
        }
        const newValue = type !== 'enable';
        const fieldSchema = effectSchema.disabled;
        const defaultValue = fieldSchema && fieldSchema.type === 'boolean'
            ? JSON.stringify(fieldSchema.default)
            : null;
        (0, save_effect_prop_1.saveEffectProp)({
            fileName: validatedLocation.source,
            nodePath,
            effectIndex,
            fieldKey: 'disabled',
            value: newValue,
            defaultValue,
            schema: effectSchema,
            setCodeValues,
        });
    }, [
        canToggle,
        effectIndex,
        effectSchema,
        nodePath,
        setCodeValues,
        validatedLocation.source,
    ]);
    const isExpanded = getIsExpanded(nodePathInfo);
    const mergedStyle = (0, react_1.useMemo)(() => ({ ...groupRowBase, ...style }), [style]);
    return (jsx_runtime_1.jsxs("div", { style: mergedStyle, children: [
            jsx_runtime_1.jsx(Padder_1.Padder, { depth: nestedDepth + 1 }), canToggle ? (jsx_runtime_1.jsx(TimelineLayerEye_1.TimelineLayerEye, { type: "effect", hidden: isDisabled, onInvoked: onToggle })) : (jsx_runtime_1.jsx(TimelineLayerEye_1.TimelineLayerEyeSpacer, {})), jsx_runtime_1.jsx(TimelineExpandArrowButton_1.TimelineExpandArrowButton, { isExpanded: isExpanded, onClick: () => toggleTrack(nodePathInfo), label: `${label} section`, disabled: false }), jsx_runtime_1.jsx("span", { style: rowLabel, children: label })
        ] }));
};
exports.TimelineEffectGroupRow = TimelineEffectGroupRow;
