// @remotion/studio/dist/components/Timeline/TimelineEffectFieldRow.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineEffectFieldRow = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const studio_shared_1 = require("@remotion/studio-shared");
const react_1 = require("react");
const remotion_1 = require("remotion");
const timeline_layout_1 = require("../../helpers/timeline-layout");
const call_api_1 = require("../call-api");
const Padder_1 = require("./Padder");
const save_prop_queue_1 = require("./save-prop-queue");
const TimelineSchemaField_1 = require("./TimelineSchemaField");
const fieldRowBase = {
    display: 'flex',
    alignItems: 'center',
    gap: 8,
    paddingRight: timeline_layout_1.EXPANDED_SECTION_PADDING_RIGHT,
};
const fieldName = {
    fontSize: 12,
    color: 'rgba(255, 255, 255, 0.8)',
    userSelect: 'none',
};
const fieldLabelRow = {
    flex: '0 0 50%',
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    gap: 6,
};
const Value = ({ field, nodePath, validatedLocation }) => {
    var _a;
    var _b;
    const { setEffectDragOverrides, clearEffectDragOverrides, setCodeValues } = (0, react_1.useContext)(remotion_1.Internals.VisualModeSettersContext);
    const { getEffectDragOverrides } = (0, react_1.useContext)(remotion_1.Internals.VisualModeDragOverridesContext);
    const { codeValues: visualModeCodeValues } = (0, react_1.useContext)(remotion_1.Internals.VisualModeCodeValuesContext);
    const effectStatus = remotion_1.Internals.getEffectCodeValuesCtx({
        codeValues: visualModeCodeValues,
        nodePath,
        effectIndex: field.effectIndex,
    });
    const propStatus = effectStatus.type === 'can-update-effect'
        ? ((_b = (_a = effectStatus.props) === null || _a === void 0 ? void 0 : _a[field.key]) !== null && _b !== void 0 ? _b : null)
        : null;
    const onDragValueChange = (0, react_1.useCallback)((value) => {
        setEffectDragOverrides(nodePath, field.effectIndex, field.key, value);
    }, [setEffectDragOverrides, nodePath, field.effectIndex, field.key]);
    const onDragEnd = (0, react_1.useCallback)(() => {
        clearEffectDragOverrides(nodePath, field.effectIndex);
    }, [clearEffectDragOverrides, nodePath, field.effectIndex]);
    const dragOverrideValue = (0, react_1.useMemo)(() => {
        const overrides = getEffectDragOverrides(nodePath, field.effectIndex);
        return overrides[field.key];
    }, [getEffectDragOverrides, nodePath, field.effectIndex, field.key]);
    const onSave = (0, react_1.useCallback)((value) => {
        if (!validatedLocation) {
            return Promise.reject(new Error('Cannot save'));
        }
        if (!propStatus) {
            return Promise.reject(new Error('Cannot save'));
        }
        if (!propStatus.canUpdate) {
            return Promise.reject(new Error('Cannot save'));
        }
        const defaultValue = field.fieldSchema.default !== undefined
            ? JSON.stringify(field.fieldSchema.default)
            : null;
        const stringifiedValue = JSON.stringify(value);
        if (value === propStatus.codeValue) {
            return Promise.resolve();
        }
        if (defaultValue === stringifiedValue &&
            propStatus.codeValue === undefined) {
            return Promise.resolve();
        }
        return (0, save_prop_queue_1.enqueueSavePropChange)({
            nodePath,
            setCodeValues,
            applyOptimistic: (prev) => (0, studio_shared_1.optimisticUpdateForEffectCodeValues)({
                previous: prev,
                effectIndex: field.effectIndex,
                fieldKey: field.key,
                value,
                schema: field.effectSchema,
            }),
            apiCall: () => (0, call_api_1.callApi)('/api/save-effect-props', {
                fileName: validatedLocation.source,
                sequenceNodePath: nodePath,
                effectIndex: field.effectIndex,
                key: field.key,
                value: stringifiedValue,
                defaultValue,
                schema: field.effectSchema,
            }),
            mergeServerResponse: (prev, data) => {
                if (!prev.canUpdate) {
                    return prev;
                }
                const idx = prev.effects.findIndex((e) => e.effectIndex === field.effectIndex);
                if (idx === -1) {
                    return {
                        ...prev,
                        effects: [...prev.effects, data],
                    };
                }
                const nextEffects = [...prev.effects];
                nextEffects[idx] = data;
                return { ...prev, effects: nextEffects };
            },
            errorLabel: 'Could not save effect prop',
        });
    }, [
        field.effectIndex,
        field.effectSchema,
        field.fieldSchema.default,
        field.key,
        nodePath,
        propStatus,
        setCodeValues,
        validatedLocation,
    ]);
    if (effectStatus.type === 'cannot-update-effect') {
        if (effectStatus.reason === 'computed') {
            return jsx_runtime_1.jsx(TimelineSchemaField_1.UnsupportedStatus, { label: "computed" });
        }
        if (effectStatus.reason === 'not-call-expression') {
            return jsx_runtime_1.jsx(TimelineSchemaField_1.UnsupportedStatus, { label: "not inline" });
        }
        if (effectStatus.reason === 'not-found') {
            return jsx_runtime_1.jsx(TimelineSchemaField_1.UnsupportedStatus, { label: "not found in code" });
        }
        throw new Error(`Unsupported effect status: ${effectStatus.reason}`);
    }
    if (effectStatus.type === 'cannot-update-sequence') {
        if (effectStatus.reason === 'not-found') {
            return jsx_runtime_1.jsx(TimelineSchemaField_1.UnsupportedStatus, { label: "not found in code" });
        }
        if (effectStatus.reason === 'error') {
            return jsx_runtime_1.jsx(TimelineSchemaField_1.UnsupportedStatus, { label: "error" });
        }
        throw new Error(`Unsupported effect status: ${effectStatus.reason}`);
    }
    if (propStatus === null || !propStatus.canUpdate) {
        return null;
    }
    const effectiveValue = remotion_1.Internals.getEffectiveVisualModeValue({
        codeValue: propStatus,
        dragOverrideValue,
        defaultValue: field.fieldSchema.default,
        shouldResortToDefaultValueIfUndefined: true,
    });
    return (jsx_runtime_1.jsx(TimelineSchemaField_1.TimelineFieldValue, { field: field, propStatus: propStatus, onSave: onSave, onDragValueChange: onDragValueChange, onDragEnd: onDragEnd, effectiveValue: effectiveValue }));
};
const TimelineEffectFieldRow = ({ field, validatedLocation, paddingLeft, nestedDepth, nodePath }) => {
    var _a;
    const style = (0, react_1.useMemo)(() => {
        return {
            ...fieldRowBase,
            height: field.rowHeight,
            paddingLeft,
        };
    }, [field.rowHeight, paddingLeft]);
    return (jsx_runtime_1.jsxs("div", { style: style, children: [
            jsx_runtime_1.jsx(Padder_1.Padder, { depth: nestedDepth + 1 }), jsx_runtime_1.jsx("div", { style: fieldLabelRow, children: jsx_runtime_1.jsx("span", { style: fieldName, children: (_a = field.description) !== null && _a !== void 0 ? _a : field.key }) }), jsx_runtime_1.jsx(Value, { field: field, nodePath: nodePath, validatedLocation: validatedLocation })
        ] }));
};
exports.TimelineEffectFieldRow = TimelineEffectFieldRow;
