// @remotion/studio-server/dist/preview-server/routes/save-effect-props.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.saveEffectPropsHandler = void 0;
const node_fs_1 = require("node:fs");
const node_path_1 = __importDefault(require("node:path"));
const renderer_1 = require("@remotion/renderer");
const studio_shared_1 = require("@remotion/studio-shared");
const parse_ast_1 = require("../../codemods/parse-ast");
const update_effect_props_1 = require("../../codemods/update-effect-props/update-effect-props");
const file_watcher_1 = require("../../file-watcher");
const undo_stack_1 = require("../undo-stack");
const watch_ignore_next_change_1 = require("../watch-ignore-next-change");
const can_update_effect_props_1 = require("./can-update-effect-props");
const can_update_sequence_props_1 = require("./can-update-sequence-props");
const format_effect_prop_change_1 = require("./log-updates/format-effect-prop-change");
const log_effect_update_1 = require("./log-updates/log-effect-update");
const log_update_1 = require("./log-updates/log-update");
const save_props_mutex_1 = require("./save-props-mutex");
const saveEffectPropsHandler = ({ input: { fileName, sequenceNodePath, effectIndex, key, value, defaultValue, schema, }, remotionRoot, logLevel, }) => (0, save_props_mutex_1.withSavePropsLock)(async () => {
    renderer_1.RenderInternals.Log.trace({ indent: false, logLevel }, `[save-effect-props] Received request for fileName="${fileName}" effectIndex=${effectIndex} key="${key}"`);
    const absolutePath = node_path_1.default.resolve(remotionRoot, fileName);
    const fileRelativeToRoot = node_path_1.default.relative(remotionRoot, absolutePath);
    if (fileRelativeToRoot.startsWith('..')) {
        throw new Error('Cannot modify a file outside the project');
    }
    const fileContents = (0, node_fs_1.readFileSync)(absolutePath, 'utf-8');
    const parsedDefault = defaultValue !== null ? JSON.parse(defaultValue) : null;
    const { output, oldValueString, formatted, logLine, effectCallee } = await (0, update_effect_props_1.updateEffectProps)({
        input: fileContents,
        sequenceNodePath: sequenceNodePath.nodePath,
        effectIndex,
        update: {
            key,
            value: JSON.parse(value),
            defaultValue: parsedDefault,
        },
    });
    const defaultValueString = parsedDefault !== null ? JSON.stringify(parsedDefault) : null;
    const normalizedOld = (0, log_update_1.normalizeQuotes)(oldValueString);
    const normalizedNew = (0, log_update_1.normalizeQuotes)(value);
    const normalizedDefault = defaultValueString !== null ? (0, log_update_1.normalizeQuotes)(defaultValueString) : null;
    const undoPropChange = (0, format_effect_prop_change_1.formatEffectPropChange)({
        effectName: effectCallee,
        key,
        oldValueString: normalizedNew,
        newValueString: normalizedOld,
        defaultValueString: normalizedDefault,
        removedProps: [],
        addedProps: [],
    });
    const redoPropChange = (0, format_effect_prop_change_1.formatEffectPropChange)({
        effectName: effectCallee,
        key,
        oldValueString: normalizedOld,
        newValueString: normalizedNew,
        defaultValueString: normalizedDefault,
        removedProps: [],
        addedProps: [],
    });
    (0, undo_stack_1.pushToUndoStack)({
        filePath: absolutePath,
        oldContents: fileContents,
        logLevel,
        remotionRoot,
        logLine,
        description: {
            undoMessage: `↩️  ${undoPropChange}`,
            redoMessage: `↪️  ${redoPropChange}`,
        },
        entryType: 'effect-props',
        suppressHmrOnFileRestore: true,
    });
    (0, undo_stack_1.suppressUndoStackInvalidation)(absolutePath);
    (0, watch_ignore_next_change_1.suppressBundlerUpdateForFile)(absolutePath);
    (0, file_watcher_1.writeFileAndNotifyFileWatchers)(absolutePath, output);
    (0, log_effect_update_1.logEffectUpdate)({
        fileRelativeToRoot,
        line: logLine,
        effectName: effectCallee,
        propKey: key,
        oldValueString,
        newValueString: value,
        defaultValueString,
        formatted,
        logLevel,
        removedProps: [],
        addedProps: [],
    });
    (0, undo_stack_1.printUndoHint)(logLevel);
    const ast = (0, parse_ast_1.parseAst)((0, node_fs_1.readFileSync)(absolutePath, 'utf-8'));
    const jsx = (0, can_update_sequence_props_1.findJsxElementAtNodePath)(ast, sequenceNodePath.nodePath);
    if (!jsx) {
        return {
            canUpdate: false,
            effectIndex,
            reason: 'not-found',
        };
    }
    return (0, can_update_effect_props_1.computeEffectPropStatus)({
        jsx,
        effectIndex,
        keys: (0, studio_shared_1.getAllSchemaKeys)(schema),
    });
});
exports.saveEffectPropsHandler = saveEffectPropsHandler;
